### PR TITLE
More specific warning for HTTPS requests on HTTP endpoints

### PR DESCRIPTION
**Pull request by raboof**
0x16 signals the start of the handshake, and produced the
expected error message on Firefox.

See also e.g. http://www.moserware.com/2009/06/first-few-milliseconds-of-https.html

As briefly discussed in #1458

### CODE REVIEW

**Comment by raboof** [comment]
Hmm, added a test but it's producing some dead letters - can you spot why?

File: akka-http-core/src/test/scala/akka/http/impl/engine/parsing/HttpsConnectionToHttpServerSpec.scala [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import akka.actor.ActorSystem
+import akka.stream.ActorMaterializer
+import akka.stream.scaladsl._
+import akka.http.scaladsl._
+import akka.http.scaladsl.model._
+
+import org.scalatest._
+import akka.testkit._
+
+class HttpsConnectionToHttpServerSpec extends AkkaSpec() {
```
Comment (jrudolph) [inline akka-http-core/src/test/scala/akka/http/impl/engine/parsing/HttpsConnectionToHttpServerSpec.scala]: I think this test would fit into ClientServerSpec? WDYT?

File: akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala [hunk 2]
```diff
@@ -694,6 +698,17 @@ class ClientServerSpec extends WordSpec with Matchers with BeforeAndAfterAll wit
...
+    "produce a useful error message when connecting to a HTTP endpoint over HTTPS" in Utils.assertAllStagesStopped {
+      val dummyFlow = Flow.fromFunction((_: HttpRequest) ⇒ ???)
+
+      val binding = Http().bindAndHandle(dummyFlow, "127.0.0.1", port = 0).futureValue
+      val uri = "https://" + binding.localAddress.getHostString + ":" + binding.localAddress.getPort
+
+      EventFilter.warning(pattern = "Perhaps this was an HTTPS request sent to an HTTP endpoint", occurrences = 6) intercept {
+        Await.ready(Http().singleRequest(HttpRequest(uri = uri)), 30.seconds)
+      }
+    }
```
Comment (jrudolph) [inline akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala]: Did you remove the `unbind` from the previous version accidentally or by purpose?

Comment (raboof) [reply akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala]: oops, no (removed it when trying to use `TestSetup`, but then decided this is clearer and forgot to put it back)

**Comment by jrudolph** [comment]
The new test failed in the nightly: https://jenkins.akka.io:8498/view/Akka-http/job/akka-http-nightly-against-akka-master-on-2.11/458/testReport/junit/akka.http.scaladsl/ClientServerSpec/The_low_level_HTTP_infrastructure_should_produce_a_useful_error_message_when_connecting_to_a_HTTP_endpoint_over_HTTPS/

File: akka-http-core/src/main/scala/akka/http/impl/engine/parsing/HttpRequestParser.scala [hunk 3]
```diff
@@ -92,10 +92,17 @@ private[http] final class HttpRequestParser(
...
-        } else throw new ParsingException(
-          BadRequest,
-          ErrorInfo("Unsupported HTTP method", s"HTTP method too long (started with '${sb.toString}'). " +
-            "Increase `akka.http.server.parsing.max-method-length` to support HTTP methods with more characters."))
+        } else {
+          if (sb.length > 0 && sb.charAt(0) == 0x16)
+            throw new ParsingException(
+              BadRequest,
+              ErrorInfo("Unsupported HTTP method", s"The HTTP method started with 0x16 rather than any known HTTP method. " +
+                "Perhaps this was an HTTPS request sent to an HTTP endpoint?"))
```
Comment (Synesso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/parsing/HttpRequestParser.scala]: I just got this message in the wild and it was immediately super helpful. Thank you!

**Pull request merged**